File: cookbook/copilot-sdk/dotnet/recipe/pr-visualization.cs
#:package GitHub.Copilot.SDK@*
#:property PublishAot=false

using System.Diagnostics;
using GitHub.Copilot.SDK;

// ============================================================================ 
// Git 與 GitHub 偵測
// ============================================================================ 

bool IsGitRepo()
{
    try
    {
        var proc = Process.Start(new ProcessStartInfo
        {
            FileName = "git",
            Arguments = "rev-parse --git-dir",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });
        proc?.WaitForExit();
        return proc?.ExitCode == 0;
    }
    catch
    {
        return false;
    }
}

string? GetGitHubRemote()
{
    try
    {
        var proc = Process.Start(new ProcessStartInfo
        {
            FileName = "git",
            Arguments = "remote get-url origin",
            RedirectStandardOutput = true,
            UseShellExecute = false,
            CreateNoWindow = true
        });

        var remoteUrl = proc?.StandardOutput.ReadToEnd().Trim();
        proc?.WaitForExit();

        if (string.IsNullOrEmpty(remoteUrl)) return null;

        // 處理 SSH: git@github.com:owner/repo.git
        var sshMatch = System.Text.RegularExpressions.Regex.Match(
            remoteUrl, @"git@github\.com:(.+/.+?)(?:\.git)?$");
        if (sshMatch.Success) return sshMatch.Groups[1].Value;

        // 處理 HTTPS: https://github.com/owner/repo.git
        var httpsMatch = System.Text.RegularExpressions.Regex.Match(
            remoteUrl, @"https://github\.com/(.+/.+?)(?:\.git)?$");
        if (httpsMatch.Success) return httpsMatch.Groups[1].Value;

        return null;
    }
    catch
    {
        return null;
    }
}

string? ParseRepoArg(string[] args)
{
    var repoIndex = Array.IndexOf(args, "--repo");
    if (repoIndex != -1 && repoIndex + 1 < args.Length)
    {
        return args[repoIndex + 1];
    }
    return null;
}

string PromptForRepo()
{
    Console.Write("輸入 GitHub 存放庫 (擁有者/存放庫): ");
    return Console.ReadLine()?.Trim() ?? "";
}

// ============================================================================ 
// 主要應用程式
// ============================================================================ 

Console.WriteLine("🔍 PR 建立時間圖表產生器\n");

// 確定存放庫
var repo = ParseRepoArg(args);

if (!string.IsNullOrEmpty(repo))
{
    Console.WriteLine($"📦 使用指定的存放庫: {repo}");
}
else if (IsGitRepo())
{
    var detected = GetGitHubRemote();
    if (detected != null)
    {
        repo = detected;
        Console.WriteLine($"📦 偵測到 GitHub 存放庫: {repo}");
    }
    else
    {
        Console.WriteLine("⚠️  找到 Git 存放庫但未偵測到 GitHub 遠端。");
        repo = PromptForRepo();
    }
}
else
{
    Console.WriteLine("📁 不在 Git 存放庫中。");
    repo = PromptForRepo();
}

if (string.IsNullOrEmpty(repo) || !repo.Contains('/'))
{
    Console.WriteLine("❌ 存放庫格式無效。應為: 擁有者/存放庫");
    return;
}

var parts = repo.Split('/');
var owner = parts[0];
var repoName = parts[1];

// 建立 Copilot 用戶端 - 不需要自訂工具！
await using var client = new CopilotClient(new CopilotClientOptions { LogLevel = "error" });
await client.StartAsync();

var session = await client.CreateSessionAsync(new SessionConfig
{
    Model = "gpt-5",
    SystemMessage = new SystemMessageConfig
    {
        Content = $"""
<context>
您正在分析 GitHub 存放庫的提取要求 (Pull Request): {owner}/{repoName}
目前工作目錄為: {Environment.CurrentDirectory}
</context>

<instructions>
- 使用 GitHub MCP Server 工具獲取 PR 資料
- 使用您的檔案和程式碼執行工具產生圖表
- 將產生的任何影像儲存到目前工作目錄
- 回應請保持簡潔
</instructions>
"""
    }
});

// 設定事件處理
session.On(evt =>
{
    switch (evt)
    {
        case AssistantMessageEvent msg:
            Console.WriteLine($"\n🤖 {msg.Data.Content}\n");
            break;
        case ToolExecutionStartEvent toolStart:
            Console.WriteLine($"  ⚙️  {toolStart.Data.ToolName}");
            break;
    }
});

// 初始提示 - 讓 Copilot 處理細節
Console.WriteLine("\n📊 開始分析...\n");

await session.SendAsync(new MessageOptions
{
    Prompt = $"""
      獲取過去一週 {owner}/{repoName} 的開放提取要求。
      計算每個 PR 的建立天數。
      然後產生一張長條圖影像，顯示 PR 建立時間的分佈
      （將它們分成合理的分組，例如 <1 天、1-3 天等）。
      將圖表儲存為目前目錄下的 "pr-age-chart.png"。
      最後，摘要 PR 健康狀況 - 平均建立時間、最舊的 PR，以及有多少可能被視為停滯。
    """
});

// 互動迴圈
Console.WriteLine("\n💡 詢問後續問題或輸入 \"exit\" 結束。\n");
Console.WriteLine("範例：");
Console.WriteLine("  - \"擴展到上個月\"");
Console.WriteLine("  - \"顯示最舊的 5 個 PR\"");
Console.WriteLine("  - \"改為產生圓餅圖\"");
Console.WriteLine("  - \"改依作者分組而非依建立時間\"");
Console.WriteLine();

while (true)
{
    Console.Write("您: ");
    var input = Console.ReadLine()?.Trim();

    if (string.IsNullOrEmpty(input)) continue;
    if (input.ToLower() is "exit" or "quit")
    {
        Console.WriteLine("👋 再見！");
        break;
    }

    await session.SendAsync(new MessageOptions { Prompt = input });
}

File: cookbook/copilot-sdk/java/recipe/PRVisualization.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS com.github:copilot-sdk-java:0.2.1-java.1

import com.github.copilot.sdk.CopilotClient;
import com.github.copilot.sdk.events.AssistantMessageEvent;
import com.github.copilot.sdk.events.ToolExecutionStartEvent;
import com.github.copilot.sdk.json.MessageOptions;
import com.github.copilot.sdk.json.PermissionHandler;
import com.github.copilot.sdk.json.SessionConfig;
import com.github.copilot.sdk.json.SystemMessageConfig;
import java.io.BufferedReader;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.regex.Pattern;

public class PRVisualization {

    public static void main(String[] args) throws Exception {
        System.out.println("🔍 PR 年齡圖表產生器\n");

        // 判斷儲存庫
        String repo;
        if (args.length > 0) {
            repo = args[0];
            System.out.println("📦 使用指定的儲存庫: " + repo);
        } else if (isGitRepo()) {
            String detected = getGitHubRemote();
            if (detected != null && !detected.isEmpty()) {
                repo = detected;
                System.out.println("📦 偵測到的 GitHub 儲存庫: " + repo);
            } else {
                System.out.println("⚠️  找到 Git 儲存庫但未偵測到 GitHub 遠端。");
                repo = promptForRepo();
            }
        } else {
            System.out.println("📁 不在 git 儲存庫中。");
            repo = promptForRepo();
        }

        if (repo == null || !repo.contains("/")) {
            System.err.println("❌ 無效的儲存庫格式。預期為: owner/repo");
            System.exit(1);
        }

        String[] parts = repo.split("/", 2);
        String owner = parts[0];
        String repoName = parts[1];

        // 建立 Copilot 用戶端
        try (var client = new CopilotClient()) {
            client.start().get();

            String cwd = System.getProperty("user.dir");
            var systemMessage = String.format("""
                <context>
                您正在分析 GitHub 儲存庫的 Pull Request: %s/%s
                目前的作業目錄為: %s
                </context>

                <instructions>
                - 使用 GitHub MCP Server 工具來擷取 PR 資料
                - 使用您的檔案和程式碼執行工具來產生圖表
                - 將任何產生的影像儲存至目前的作業目錄
                - 回覆請簡潔
                </instructions>
                """, owner, repoName, cwd);

            var session = client.createSession(
                new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)
                    .setModel("gpt-5")
                    .setSystemMessage(new SystemMessageConfig().setContent(systemMessage))
            ).get();

            // 設定事件處理
            session.on(AssistantMessageEvent.class, msg -> 
                System.out.println("\n🤖 " + msg.getData().content() + "\n")
            );

            session.on(ToolExecutionStartEvent.class, evt -> 
                System.out.println("  ⚙️  " + evt.getData().toolName())
            );

            // 初始提示 - 讓 Copilot 自行判斷詳細資料
            System.out.println("\n📊 開始分析...\n");

            String prompt = String.format("""
                擷取 %s/%s 上週以來開啟的 pull requests。
                計算每個 PR 的年齡 (以天數計)。
                然後產生顯示 PR 年齡分佈的長條圖影像
                (將它們分組至合理的區間，例如 <1 天、1-3 天等)。
                將圖表以 "pr-age-chart.png" 為名儲存於目前目錄。
                最後，總結 PR 品質 - 平均年齡、最舊的 PR，以及多少可能被視為已過時。
                """, owner, repoName);

            session.sendAndWait(new MessageOptions().setPrompt(prompt)).get();

            // 互動式迴圈
            System.out.println("\n💡 詢問後續問題或輸入 \"exit\" 以退出。\n");
            System.out.println("範例:");
            System.out.println("  - \"擴展至上個月\"");
            System.out.println("  - \"顯示我 5 個最舊的 PR\"");
            System.out.println("  - \"改為產生圓餅圖\"");
            System.out.println("  - \"依作者而非年齡分組\"");
            System.out.println();

            try (var reader = new BufferedReader(new InputStreamReader(System.in))) {
                while (true) {
                    System.out.print("您: ");
                    String input = reader.readLine();
                    if (input == null) break;
                    input = input.trim();

                    if (input.isEmpty()) continue;
                    if (input.equalsIgnoreCase("exit") || input.equalsIgnoreCase("quit")) {
                        System.out.println("👋 再見!");
                        break;
                    }

                    session.sendAndWait(new MessageOptions().setPrompt(input)).get();
                }
            }

            session.close();
        }
    }

    // ============================================================================
    // Git 與 GitHub 偵測
    // ============================================================================

    private static boolean isGitRepo() {
        try {
            Process proc = Runtime.getRuntime().exec(new String[]{"git", "rev-parse", "--git-dir"});
            return proc.waitFor() == 0;
        } catch (Exception e) {
            return false;
        }
    }

    private static String getGitHubRemote() {
        try {
            Process proc = Runtime.getRuntime().exec(new String[]{"git", "remote", "get-url", "origin"});
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(proc.getInputStream()))) {
                String remoteURL = reader.readLine();
                if (remoteURL == null) return null;
                remoteURL = remoteURL.trim();

                // 處理 SSH: git@github.com:owner/repo.git
                var sshPattern = Pattern.compile("git@github\\.com:(.+/.+?)(?:\\.git)?$");
                var sshMatcher = sshPattern.matcher(remoteURL);
                if (sshMatcher.find()) {
                    return sshMatcher.group(1);
                }

                // 處理 HTTPS: https://github.com/owner/repo.git
                var httpsPattern = Pattern.compile("https://github\\.com/(.+/.+?)(?:\\.git)?$");
                var httpsMatcher = httpsPattern.matcher(remoteURL);
                if (httpsMatcher.find()) {
                    return httpsMatcher.group(1);
                }
            }
        } catch (Exception e) {
            // 忽略
        }
        return null;
    }

    private static String promptForRepo() throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.print("輸入 GitHub 儲存庫 (owner/repo): ");
        String line = reader.readLine();
        if (line == null) {
            throw new EOFException("讀取儲存庫名稱時輸入結束");
        }
        return line.trim();
    }
}
